### ISSUE TITLE
How to use cvat-sdk with token

### ISSUE THREAD

**Issue by liudaolunboluo** [cvat-ai/cvat#7439]
### Actions before raising this issue

- [X] I searched the existing issues and did not find anything similar.
- [X] I read/searched [the docs](https://github.com/cvat-ai/cvat/tree/master#documentation)

### Is your feature request related to a problem? Please describe.

When creating a client in cvat-sdk you need to enter the username and password but now I have multiple users and I don't know their passwords but I can get their tokens, can I use this token directly to use the cvat-sdk client?

### Describe the solution you'd like

I us token directly to use the cvat-sdk client

**Comment by zhiltsov-max** [comment]
Hi, there is no "official way", but please check [this answer](https://github.com/opencv/cvat/issues/6736#issuecomment-1689459971) and [this draft PR](https://github.com/opencv/cvat/pull/7320).

**Comment by liudaolunboluo** [comment]
> Hi, there is no "official way", but please check [this answer](https://github.com/opencv/cvat/issues/6736#issuecomment-1689459971) and [this draft PR](https://github.com/opencv/cvat/pull/7320).

Are sessionid and csrftoken necessary? Because my scenario is that I deployed a privatized cvat myself and integrated him into my system, and I want to automate the import of the dataset in one of my system's features, so I shouldn't be able to get the cross-domain cookie information in the browser

**Comment by liudaolunboluo** [comment]
> If you're working as a local admin, you should be able to download annotations from other users. Consider using this approach instead. Would a special service account for these purposes work for you?

I only want to upload dataset to project,now I think  I can use normal user create project and use admin user upload datset,because I have admin user password,but question is high level sdk have not only import dataset api,it only have create project from datset api. so what can i do?

**Comment by liudaolunboluo** [comment]
> If you're working as a local admin, you should be able to download annotations from other users. Consider using this approach instead. Would a special service account for these purposes work for you?

oh,sory,I read this manual: https://opencv.github.io/cvat/docs/api_sdk/sdk/highlevel-api/，That's very helpful.,
it worked:
project = client.projects.retrieve(project_id)

**Comment by liudaolunboluo** [comment]
I'm having a little problem again.
My code:
```python
with make_client(
        host="my local server url",
        credentials=('admin', 'admin password')

) as client:
    try:
        project = client.projects.retrieve(project_id)
    except Exception as e:
        print('throw exception：', e)
    project.import_dataset(format_name='CVAT 1.1', filename='dataset file path', pbar=pbar)
```
it throw exception:
HTTP response body: b'{"detail":"CSRF Failed: CSRF token missing or incorrect."}'
I've noticed that in some issues uploading on the page may encounter this problem, just try logging out and logging in, now in the sdk I'm still encountering this problem, how can I fix it? Thank you very much for your reply!

**Comment by liudaolunboluo** [comment]
I solved it! It looks like it was a code issue, there was no X-Csrftoken set for the header anywhere in the code for importing the dataset, so it was reporting an error：https://stackoverflow.com/questions/26639169/csrf-failed-csrf-token-missing-or-incorrect
,I'm now changing it this way and it doesn't report an error:
```python
        pairs = client.api_client.get_common_headers()['Cookie'].split('; ')
        dictionary = {pair.split('=')[0]: pair.split('=')[1] for pair in pairs}
        client.api_client.set_default_header("X-Csrftoken", dictionary['csrftoken'])
        client.api_client.get_common_headers()
```
The core is to get the csrftoken from the cookie and set the X-CSRFToken to the header.
but But there is a new error reported:
```
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/cvat_sdk/core/proxies/projects.py", line 57, in import_dataset
    DatasetUploader(self._client).upload_file_and_wait(
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/cvat_sdk/core/uploading.py", line 321, in upload_file_and_wait
    rq_id = json.loads(response.data).get("rq_id")
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
code is:rq_id = json.loads(response.data).get("rq_id")
But response.data doesn't look like a json string but a byte array

**Comment by liudaolunboluo** [comment]
I solved it again,But the sdk api does have problems too, hopefully it will be fixed soon!

```python
project = client.projects.retrieve(project_id)
filename = Path(dataset_path)
params = {"format": 'CVAT 1.1', "filename": filename.name}
url = client.api_map.make_endpoint_url(project.api.create_dataset_endpoint.path, kwsub={"id": project_id})
response = DatasetUploader(client).upload_file(
                url, filename, pbar=pbar, query_params=params, meta={"filename": params["filename"]})
```

**Comment by zhiltsov-max** [comment]
Thank you for reporting the issues. 

> But response.data doesn't look like a json string but a byte array

Probably, you're using an older version of CVAT, before the change was introduced. Please make sure SDK and server versions match. The change was introduced in https://github.com/opencv/cvat/pull/5909.

**Comment by Abo-Omar-74** [comment]
I am looking forward to contributing to solving this issue through GSOC 24 and I have some questions regarding the scope of [this](https://github.com/opencv/cvat/wiki/GSoC-2024#idea-API-keys-and-token-based-auth-for-sdk-and-cli) project. 

- **Scope Clarification:**
    
    - According to the [documentation](https://opencv.github.io/cvat/docs/api_sdk/sdk/lowlevel-api/#authentication), there is already a way of Authentication using tokens. Does this mean that the scope of this problem is to make the process of creating these tokens easier with the UI?
    
- **Integration with `auth_api`:**
    
    - Do I need to add a new method to `auth_api` for token generation, or utilize the existing login method since it returns an access token?
    
- **Display of Access Tokens:**
    
    - Should the access token be directly visible in the account settings, allowing users to copy and manually include it in the request headers using `api_client.set_default_header("Authorization", "Token " + {generated token})`?
    - Alternatively, should the UI indicate successful addition of the access token and automate its inclusion in environment variables for seamless integration into request headers?

**Comment by Abo-Omar-74** [comment]
> I am looking forward to contributing to solving this issue through GSOC 24 and I have some questions regarding the scope of [this](https://github.com/opencv/cvat/wiki/GSoC-2024#idea-API-keys-and-token-based-auth-for-sdk-and-cli) project.
> 
> * **Scope Clarification:**
>   
>   * According to the [documentation](https://opencv.github.io/cvat/docs/api_sdk/sdk/lowlevel-api/#authentication), there is already a way of Authentication using tokens. Does this mean that the scope of this problem is to make the process of creating these tokens easier with the UI?
> * **Integration with `auth_api`:**
>   
>   * Do I need to add a new method to `auth_api` for token generation, or utilize the existing login method since it returns an access token?
> * **Display of Access Tokens:**
>   
>   * Should the access token be directly visible in the account settings, allowing users to copy and manually include it in the request headers using `api_client.set_default_header("Authorization", "Token " + {generated token})`?
>   * Alternatively, should the UI indicate successful addition of the access token and automate its inclusion in environment variables for seamless integration into request headers?

I would greatly appreciate your guidance in clarifying this. Your expertise would be incredibly valuable. I truly appreciate your support.

@zhiltsov-max @SpecLad @azhavoro

**Comment by zhiltsov-max** [comment]
@Abo-Omar-74,

Hi, thank you for reaching us out about this topic. Overall, the scope for GSoC can vary, depending on your skills and desire. This is a complex task, that can be split into several elements:
1. SDK / CLI update to support persistent login. Can be done with any type of keys. Basically, it means integrating this code snippet into our current CLI:

<details>

This code allows to record user credentials in the user profile directory. Credentials for each visited CVAT host are recorded, to allow future visits without explicit authentication from the user.

```python
from __future__ import annotations

import json
import os
from argparse import ArgumentParser
from functools import partial
from pathlib import Path
from types import SimpleNamespace
from typing import Optional, Tuple

import attrs
import cvat_sdk

API_KEY_VAR = "CVAT_API_KEY"
API_SESSIONID_VAR = "CVAT_API_SESSIONID"
API_CSRFTOKEN_VAR = "CVAT_API_CSRFTOKEN"


def add_cli_parser_args(parser: ArgumentParser) -> ArgumentParser:
    parser.add_argument("--org")
    parser.add_argument("--host", default="https://app.cvat.ai")
    parser.add_argument("--port")
    parser.add_argument(
        "--login",
        help=f"A 'login:password' pair. "
        f"Default: use {API_KEY_VAR}, {API_SESSIONID_VAR}, {API_CSRFTOKEN_VAR} env vars",
    )
    parser.add_argument("--profile-dir", default=None, type=Path, help="User profile dir")

    return parser


DEFAULT_PROFILE_FILENAME = "profile.json"
DEFAULT_PROFILE_DIR = "~/cvat/"


@attrs.define
class UserCredentials:
    token: str
    sessionid: str
    csrftoken: str


@attrs.define
class UserProfile:
    _PROFILE_UMASK = 0o600

    @classmethod
    def get_default_path(cls) -> Path:
        return Path(DEFAULT_PROFILE_DIR).expanduser() / DEFAULT_PROFILE_FILENAME

    path: Optional[Path] = attrs.field(factory=partial(get_default_path.__func__, __build_class__))

    credentials: dict[str, UserCredentials] = attrs.field(factory=dict)
    extra: dict[str, str] = attrs.field(factory=dict)

    @classmethod
    def parse(cls, path: Path) -> UserProfile:
        data = json.loads(path.read_text())

        return UserProfile(
            credentials={k: UserCredentials(**v) for k, v in data["credentials"].items()},
            extra=data.get("extra"),
        )

    @classmethod
    def load(cls, path: Optional[Path] = None) -> UserProfile:
        path = path or cls.get_default_path()

        if path.is_file():
            if (mode := path.stat().st_mode) & cls._PROFILE_UMASK != cls._PROFILE_UMASK:
                raise Exception(f"Invalid profile mode. Expected 600 (rw--), got {oct(mode)}")

            profile = cls.parse(path)
        else:
            profile = UserProfile(path=path)
            profile.save(path)

        return profile

    def save(self, path: Optional[Path] = None):
        path = path or self.path or self.get_default_path()

        data = json.dumps(
            attrs.asdict(self, filter=lambda a, v: a.name != "path", recurse=True), indent=2
        )

        if path.absolute() == self.get_default_path().absolute() and not path.parent.is_dir():
            path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)

        with path.open("w", encoding="utf-8") as f:
            f.write(data)

        path.chmod(self._PROFILE_UMASK)

    def update_credentials(self, host: str, *, token: str, sessionid: str, csrftoken: str):
        self.credentials[self._make_host_key(host)] = UserCredentials(
            token=token, sessionid=sessionid, csrftoken=csrftoken
        )

    def has_credentials(self, host: str) -> bool:
        return self._make_host_key(host) in self.credentials

    def get_credentials(self, host: str) -> UserCredentials:
        return self.credentials[self._make_host_key(host)]

    def _make_host_key(self, host: str) -> str:
        return host.split("://", maxsplit=1)[-1]


class UserClient(cvat_sdk.Client):
    def __init__(self, *args, profile: Optional[UserProfile] = None, **kwargs) -> None:
        super().__init__(*args, **kwargs)

        self.profile = profile or UserProfile.load()

    def login(self, credentials: Tuple[str, str]) -> None:
        super().login(credentials)
        self.save_current_credentials()

    def load_credentials(self):
        credentials = self.profile.get_credentials(self.api_map.host)

        self.api_client.set_default_header("Authorization", f"Token {credentials.token}")
        self.api_client.cookies["sessionid"] = credentials.sessionid
        self.api_client.cookies["csrftoken"] = credentials.csrftoken
        self.api_client.set_default_header("X-Csrftoken", credentials.csrftoken)

    def save_current_credentials(self):
        self.profile.update_credentials(
            host=self.api_map.host,
            token=self.api_client.default_headers["Authorization"].split("Token ")[-1],
            sessionid=self.api_client.cookies["sessionid"].value,
            csrftoken=self.api_client.cookies["csrftoken"].value,
        )
        self.profile.save()


def make_client_from_cli(parsed_args: SimpleNamespace) -> UserClient:
    profile = UserProfile.load(parsed_args.profile_dir)

    host = parsed_args.host
    port = parsed_args.port
    url = host.rstrip("/")
    if port:
        url = f"{url}:{port}"

    with UserClient(url, profile=profile) as client:
        if parsed_args.org:
            client.organization_slug = parsed_args.org

        if parsed_args.login:
            client.login(parsed_args.login.split(":", maxsplit=1))
        elif api_key := os.getenv(API_KEY_VAR):
            client.api_client.set_default_header("Authorization", f"Token {api_key}")
            client.api_client.cookies["sessionid"] = os.getenv(API_SESSIONID_VAR)
            client.api_client.cookies["csrftoken"] = os.getenv(API_CSRFTOKEN_VAR)
            client.api_client.set_default_header("X-Csrftoken", os.getenv(API_CSRFTOKEN_VAR))
            client.save_current_credentials()
        elif profile.has_credentials(client.api_map.host):
            client.load_credentials()

        return client

```

</details>

I can see it can be extended with an `auth` CLI command, that allows just to login on a host and store tokens locally for further use. It can also be extended with a command to remove any of the recorded tokens from the local profile.

1. Server updates to support manageable API token generation for a user
2. UI updates to support API token management options in the personal account page (basically, requires to add such page first, as there is no such section in CVAT yet). Probably, can work similarly to what's on GitHub.
3. SDK / CLI updates to support API tokens for auth - similar to the existing login/password pair

Basically, I'd propose to start with the first task from this list. It already has a PoC implementation, doesn't require too many changes, and just needs to be productized - i.e. with tests and convenient user interface.

**Comment by ritikraj26** [comment]
Hi, I'm interested in contributing to this project. My understanding is that the goal is to implement a system within user profiles that allows for the generation of API access tokens. Users should be able to store these tokens locally for persistent authorization, similar to how GitHub handles personal access tokens. Can you confirm if this understanding is correct? 

@zhiltsov-max @nmanovic

**Comment by zhiltsov-max** [comment]
@ritikraj26, yes, your understanding is correct.

**Comment by ritikraj26** [comment]
Currently, the user is obtaining the auth token using the username and password. We need to just replace the current method with the API access tokens generated by the user.
@zhiltsov-max

**Comment by ritikraj26** [comment]
@zhiltsov-max 
Where can I connect with you? Would you be open to some discussion? I am keenly interested in contributing to this project.

**Comment by zhiltsov-max** [comment]
@ritikraj26, I think it's best to be discussed here for others to see and participate, unless you're going to send your GSoC proposal. Proposals need to be sent via the GSoC site.

> The rest of the flow remains the same. Only the method to obtain the auth token is updated?

Basically - yes, but "The rest of the flow" needs clarification. If you meant the point 1 from [the comment](https://github.com/opencv/cvat/issues/7439#issuecomment-2002413060) above, then I think it's quite close to what's expected from the token use point of view. But tokens can also be managed in CVAT as discussed in the comment.

**Comment by ritikraj26** [comment]
Hi, I recently switched from a ubuntu to mac, I am having some trouble setting up. 
<img width="1609" alt="Screenshot 2024-10-25 at 6 13 38 PM" src="https://github.com/user-attachments/assets/7dade1ec-5cd8-48d2-9818-8b6e32247211">

Some help please. I am looking forward to complete this project.

**Comment by azhavoro** [comment]
> Hi, I recently switched from a ubuntu to mac, I am having some trouble setting up. <img alt="Screenshot 2024-10-25 at 6 13 38 PM" width="1609" src="https://private-user-images.githubusercontent.com/76646497/380165834-7dade1ec-5cd8-48d2-9818-8b6e32247211.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4NjA2ODksIm5iZiI6MTcyOTg2MDM4OSwicGF0aCI6Ii83NjY0NjQ5Ny8zODAxNjU4MzQtN2RhZGUxZWMtNWNkOC00OGQyLTk4MTgtOGI2ZTMyMjQ3MjExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI1VDEyNDYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlNWM5MzM4MTQ5MmNjMWI5OTQ3ODM3MTc2OTIzYWMyMzIzZGIzZjkwNDUzNDdmYjUxOWUxYzQyNmQ2OWUzNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Xjh9TYnbcwBBwD87JbK_44U-57ikJ7Kjsm3OVOpYnyA">
> 
> Some help please. I am looking forward to complete this project.

https://stackoverflow.com/questions/69818376/localhost5000-unavailable-in-macos-v12-monterey

**Comment by ritikraj26** [comment]
> > Hi, I recently switched from a ubuntu to mac, I am having some trouble setting up. <img alt="Screenshot 2024-10-25 at 6 13 38 PM" width="1609" src="https://private-user-images.githubusercontent.com/76646497/380165834-7dade1ec-5cd8-48d2-9818-8b6e32247211.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4NjA2ODksIm5iZiI6MTcyOTg2MDM4OSwicGF0aCI6Ii83NjY0NjQ5Ny8zODAxNjU4MzQtN2RhZGUxZWMtNWNkOC00OGQyLTk4MTgtOGI2ZTMyMjQ3MjExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI1VDEyNDYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlNWM5MzM4MTQ5MmNjMWI5OTQ3ODM3MTc2OTIzYWMyMzIzZGIzZjkwNDUzNDdmYjUxOWUxYzQyNmQ2OWUzNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Xjh9TYnbcwBBwD87JbK_44U-57ikJ7Kjsm3OVOpYnyA">
> > Some help please. I am looking forward to complete this project.
> 
> https://stackoverflow.com/questions/69818376/localhost5000-unavailable-in-macos-v12-monterey

This issue is resolved. However, I am still not able to see an requests on the debug server.

<img width="1710" alt="Screenshot 2024-10-27 at 4 02 43 PM" src="https://github.com/user-attachments/assets/accd9f37-1abc-4331-8987-f7c13a6a6075">
<img width="1710" alt="Screenshot 2024-10-27 at 4 03 20 PM" src="https://github.com/user-attachments/assets/af869311-e80f-4e2e-a130-57347968a6e6">

**Comment by zhiltsov-max** [comment]
@ritikraj26, please try the following:
- Check the server logs after you tried to open the CVAT page in the browser. If there are no requests listed in the log, try to replace `localhost` with `127.0.0.1` in `cvat-ui/package.json:8` (`"start": "webpack serve --env API_URL=http://127.0.0.1:7000 --config ./webpack.config.js --mode=development",`)
- Then you also might need to start OPA and several other containers, as shown in the development environment setup guide (`docker compose -f docker-compose.yml -f docker-compose.dev.yml up -d --build \
  cvat_opa cvat_db cvat_redis_inmem cvat_redis_ondisk cvat_server`)

**Comment by ritikraj26** [comment]
> @ritikraj26, please try the following:
> 
> * Check the server logs after you tried to open the CVAT page in the browser. If there are no requests listed in the log, try to replace `localhost` with `127.0.0.1` in `cvat-ui/package.json:8` (`"start": "webpack serve --env API_URL=http://127.0.0.1:7000 --config ./webpack.config.js --mode=development",`)
> * Then you also might need to start OPA and several other containers, as shown in the development environment setup guide (`docker compose -f docker-compose.yml -f docker-compose.dev.yml up -d --build cvat_opa cvat_db cvat_redis_inmem cvat_redis_ondisk cvat_server`)

Changing localhost to 127.0.0.1 worked. Thanks.

**Comment by zhiltsov-max** [comment]
Related: https://github.com/cvat-ai/cvat/issues/4961, https://github.com/cvat-ai/cvat/issues/8660

**Comment by ritikraj26** [comment]
@zhiltsov-max Has this project been completed?

**Comment by ritikraj26** [comment]
While installing the requirements file I am getting this issue (logs below)

Collecting datumaro@ git+https://github.com/cvat-ai/datumaro.git@265bf06a4f742235828f9f53f7f7511a10217437 (from -r /Users/ritik/Projects/cvat/cvat/requirements/base.txt (line 56))
  Cloning https://github.com/cvat-ai/datumaro.git (to revision 265bf06a4f742235828f9f53f7f7511a10217437) to /private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-install-hu4xhbap/datumaro_44ed75da54584a909896b64745a930ce
  Running command git clone --filter=blob:none --quiet https://github.com/cvat-ai/datumaro.git /private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-install-hu4xhbap/datumaro_44ed75da54584a909896b64745a930ce
  Running command git rev-parse -q --verify 'sha^265bf06a4f742235828f9f53f7f7511a10217437'
  Running command git fetch -q https://github.com/cvat-ai/datumaro.git 265bf06a4f742235828f9f53f7f7511a10217437
  Running command git checkout -q 265bf06a4f742235828f9f53f7f7511a10217437
  Resolved https://github.com/cvat-ai/datumaro.git to commit 265bf06a4f742235828f9f53f7f7511a10217437
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting av==9.2.0 (from -r /Users/ritik/Projects/cvat/utils/dataset_manifest/requirements.txt (line 6))
  Using cached av-9.2.0.tar.gz (2.4 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [81 lines of output]
      Compiling av/plane.pyx because it changed.
      [1/1] Cythonizing av/plane.pyx
      Compiling av/dictionary.pyx because it changed.
      [1/1] Cythonizing av/dictionary.pyx
      Compiling av/stream.pyx because it changed.
      [1/1] Cythonizing av/stream.pyx
      Compiling av/option.pyx because it changed.
      [1/1] Cythonizing av/option.pyx
      warning: av/enum.pyx:321:4: __nonzero__ was removed in Python 3; use __bool__ instead
      Compiling av/enum.pyx because it changed.
      [1/1] Cythonizing av/enum.pyx
      Compiling av/bytesource.pyx because it changed.
      [1/1] Cythonizing av/bytesource.pyx
      Compiling av/buffer.pyx because it changed.
      [1/1] Cythonizing av/buffer.pyx
      Compiling av/packet.pyx because it changed.
      [1/1] Cythonizing av/packet.pyx
      Compiling av/error.pyx because it changed.
      [1/1] Cythonizing av/error.pyx
      Compiling av/_core.pyx because it changed.
      [1/1] Cythonizing av/_core.pyx
      Compiling av/format.pyx because it changed.
      [1/1] Cythonizing av/format.pyx
      performance hint: av/logging.pyx:232:0: Exception check on 'log_callback' will always require the GIL to be acquired.
      Possible solutions:
          1. Declare 'log_callback' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.
          2. Use an 'int' return type on 'log_callback' to allow an error code to be returned.
      
      Error compiling Cython file:
      ------------------------------------------------------------
      ...
      cdef const char *log_context_name(void *ptr) nogil:
          cdef log_context *obj = <log_context*>ptr
          return obj.name
      
      cdef lib.AVClass log_class
      log_class.item_name = log_context_name
                            ^
      ------------------------------------------------------------
      
      av/logging.pyx:216:22: Cannot assign type 'const char *(void *) except? 0 nogil' to 'const char *(*)(void *) noexcept nogil'. Exception values are incompatible. Suggest adding 'noexcept' to the type of 'log_context_name'.
      
      Error compiling Cython file:
      ------------------------------------------------------------
      ...
      
      # Start the magic!
      # We allow the user to fully disable the logging system as it will not play
      # nicely with subinterpreters due to FFmpeg-created threads.
      if os.environ.get('PYAV_LOGGING') != 'off':
          lib.av_log_set_callback(log_callback)
                                  ^
      ------------------------------------------------------------
      
      av/logging.pyx:351:28: Cannot assign type 'void (void *, int, const char *, va_list) except * nogil' to 'av_log_callback' (alias of 'void (*)(void *, int, const char *, va_list) noexcept nogil'). Exception values are incompatible. Suggest adding 'noexcept' to the type of 'log_callback'.
      Compiling av/logging.pyx because it changed.
      [1/1] Cythonizing av/logging.pyx
      Traceback (most recent call last):
        File "/Users/ritik/Projects/cvat/.env/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
          main()
        File "/Users/ritik/Projects/cvat/.env/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 373, in main
          json_out["return_val"] = hook(**hook_input["kwargs"])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/ritik/Projects/cvat/.env/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 143, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-build-env-dwahyf2_/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 331, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-build-env-dwahyf2_/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 301, in _get_build_requires
          self.run_setup()
        File "/private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-build-env-dwahyf2_/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 512, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-build-env-dwahyf2_/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 317, in run_setup
          exec(code, locals())
        File "<string>", line 157, in <module>
        File "/private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-build-env-dwahyf2_/overlay/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1154, in cythonize
          cythonize_one(*args)
        File "/private/var/folders/0v/v8dby0qn01q8g5d3tn_h9gsh0000gn/T/pip-build-env-dwahyf2_/overlay/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1298, in cythonize_one
          raise CompileError(None, pyx_file)
      Cython.Compiler.Errors.CompileError: av/logging.pyx
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

**Comment by zhiltsov-max** [comment]
@ritikraj26, the feature is being reviewed now. There's currently no UI PR at the moment.

The error with `av` is because we're using an old version of the library now, which doesn't have pre-built binaries in PyPI at the moment (they were removed at some point). Please try to use building steps from the server `Dockerfile` for local installation or just build the server container.
